src/hotspot/share/runtime/mrswMutex.hpp

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_RUNTIME_MRSWMUTEX_HPP
+#define SHARE_RUNTIME_MRSWMUTEX_HPP
+
+#include "javaThread.hpp"
+#include "memory/allocation.hpp"
+#include "runtime/atomic.hpp"
+#include "runtime/mutex.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/macros.hpp"
+
+class NoTransition;
+
+// Multi-reader single-writer lock implementation.
+// If a transition needs to occur when the reader or writer lock is taken (in
+// case the thread blocks), then specialize on the ReaderTransition and/or
+// WriterTransition parameters. This lock is unfair, high contention of readers
+// may starve some of them.
+// Writers take precedence, blocking new readers from entering and allowing
+// current readers to proceed. The consequences are undefined if one or more
+// writer threads attempt to enter their critical region when another writer
+// thread already is attempting to do so or currently is in its critical region.
+class MRSWMutex : public CHeapObj<mtSynchronizer> {
+private:
+  NONCOPYABLE(MRSWMutex);
+  class NoTransition : public StackObj {
+  public:
+    NoTransition(Thread* thread) {
+    }
+  };
+
+  template<typename T>
+  class Locker : public StackObj {
+  private:
+    T* const _lock;
+
+  public:
+    Locker(T* lock)
+      : _lock(lock) {
+      if (_lock != nullptr) {
```

Comment:
It would make sense in the same circumstances the MutexLocker also takes a NULL argument.

Reply:
I think the pattern of passing a nullptr argument to Lockers is a pattern that we reach for quite often. The benefit is that it allows for a conditional scope, without messing up the surrounding code.

Compare
```
Locker locker(should_lock ? &_lock : nullptr);
doit1();
doit2();
doit3();
```
vs
```
if (should_lock) {
  Locker locker(should_lock ? &_lock : nullptr);
  doit1();
  doit2();
  doit3();
} else {
  doit1();
  doit2();
  doit3();
}
```
or, having to go extra miles to pull out the code into helper functions:
```
void MyClass::doall() {
  doit1();
  doit2();
  doit3();
}

if (should_lock) {
  Locker locker(should_lock ? &_lock : nullptr);
  doall();
} else {
  doall();
}
```

The first version is much easier to read, IMHO. If you want to ban nullptrs in Lockers, I'd like to see an alternative that retains the crispness of the code.

Reply:
You can have a blah_locked() version in these cases.  Having code called in multiple situations (some paths locked and some not) might be not great.  Some higher level of code should have done the locking or not.
That said, the sweeper has multiple of these with the Compile_lock so this can't be attempted to be squashed until the sweeper is gone.

Reply:
I've gotten more push-back on things called blah_locked(). The motivation was that it's not clear if it is the caller or callee who should take the lock, so that name convention isn't really helping the reader. I'm a but sympathetic to that point-of-view.

I still don't think this is something we should be squashing. Maybe it would help sway me if someone tried to convert ZActivatedArray and show what the code would look like if we couldn't use this pattern:
https://github.com/openjdk/zgc/blob/zgc_generational/src/hotspot/share/gc/z/zArray.inline.hpp

Alternatively, we could have two versions of lockers to appease both sides? One that never accepts nullptr, which would be the locker that most should be using, and one for those cases where we do need to run with and without the lock held.

Reply:
I don't now if your zlock is even a standard runtime/mutex, but let's definitely not have another version of lockers.  I've never pushed back on blah_locked().  I need to see more evidence of why we need this in the code (after the sweeper is removed), before I suggest a policy change.
I have a patch that removes the NULL ObjectLocker argument. I thought ObjectLocker was the only thing that took null until yesterday.

Reply:
> Alternatively, we could have two versions of lockers to appease both sides?

"They who cannot remember the past are condemned to repeat it." :)

We used to have `MutexLockerEx` that allowed the NULL whereas `MutexLocker` did not. 

I have no issues with the NULL case as it is cleaner that defining locked and unlocked code paths.

Reply:
MutexLockerEx did two things - allow a nullptr for the mutex argument, and
allow locking without a safepoint check. The vast majority of uses were for
the latter. Both of those got merged into MutexLocker. (The old MutexLocker
was deleted and MutexLockerEx was renamed to MutexLocker, with uses updated.)
Perhaps only the safepoint check control should have been added to
MutexLocker.